reference/SimpleCalculator/SimpleCalculator.Shared/Business/Calculator.cs

```diff
@@ -1,183 +1,238 @@
-                    };
-                }
-            }
-            else
+                "÷" => calculator.Number1!.Value / number2!.Value,
+                "×" => calculator.Number1!.Value * number2!.Value,
+                "+" => calculator.Number1!.Value + number2!.Value,
+                "−" => calculator.Number1!.Value - number2!.Value,
+                _   => throw new InvalidOperationException()
+            };
+
+            calculator = calculator with
             {
-                calculator = calculator with { Number = calculator.Number + key };
-            }
+                Number2 = number2,
+                Result = result
+            };
+        }
+
+        return calculator;
+    }
+
+    private Calculator ProcessPercentageKey(Calculator calculator)
+    {
+        if (calculator.HasOperator && calculator.HasNumber1)
+        {
+            double? number2 = calculator.HasNumber ? GetNumber(calculator.Number) : 0.0;
 
-            return calculator;
+            double result = calculator.Operator switch
+            {
+                "÷" => calculator.Number1!.Value / (number2!.Value / 100) * calculator.Number1!.Value,
+                "×" => calculator.Number1!.Value * (number2!.Value / 100) * calculator.Number1!.Value,
+                "+" => calculator.Number1!.Value + (number2!.Value / 100) * calculator.Number1!.Value,
+                "−" => calculator.Number1!.Value - (number2!.Value / 100) * calculator.Number1!.Value,
+                _   => throw new InvalidOperationException()
+            };
+
+            calculator = calculator with
+            {
+                Number2 = number2,
+                Result = result,
+                IsNumber2Percentage = true
+            };
         }
 
-        double? GetNumber(string number)
+        return calculator;
+    }
+
+    private Calculator ProcessPlusMinusKey(Calculator calculator)
+    {
+        if (calculator.HasNumber)
         {
-            return Convert.ToDouble(number);
+            calculator = calculator with { Number = calculator.Number?.StartsWith("-") == true ? calculator.Number?.Substring(1) : "-" + calculator.Number };
         }
+        return calculator;
     }
+
+    private Calculator ProcessOperatorKey(Calculator calculator, string key)
+    {
+        if (calculator.HasNumber && !calculator.HasOperator)
+        {
+            calculator = calculator with
+            {
+                Operator = key,
+                Number1 = GetNumber(calculator.Number),
+                Number = null
+            };
+        }
+
+        return calculator;
+    }
+
+    double? GetNumber (string number)
+        => Convert.ToDouble(number);
+}
+
+public enum KeyInput
+{
+    Multiplication,
+    Division,
+    Subtraction,
+    Addition,
+    Back,
+    Dot,
+    Clear,
+    Equal,
+    Percentage,
+    PlusMinus,
+    Zero,
+    One,
+    Two,
+    Three,
+    Four,
+    Five,
+    Six,
+    Seven,
+    Eight,
+    Nine
```

Comment:
save a headache and put these at the beginning of the enum like:
```cs
Zero = 0,
One = 1,
Two = 2
```
This way when it is a numeric value you can just cast to an int and append it.

Reply:
Changes in next commit